### PR TITLE
Riscv32e align stack to 4 bytes

### PR DESCRIPTION
**Pull request by cheme**
To build some riscv32 embeded binary, I had to change the stack alignment.

Not sure if this needs to be in master as no current cpu touch it (at least this PR would be accessible from github if someone else need it).

I was using 
```
    pub const pvm = CpuModel{
        .name = "pvm",
        .llvm_name = null,
        .features = featureSet(&[_]Feature{
            .@"32bit",
            .e,
            .a,
            .m,
        }),
    };
```
and `zig build -Dtarget=riscv32-freestanding -Dcpu=pvm` to produce:
```
Attribute Section: riscv
File Attributes
  Tag_RISCV_stack_align: 4-bytes
  Tag_RISCV_arch: "rv32e2p0_m2p0_a2p1"
```
But it is not a real cpu (I did not find a convenient way to pass custom features from build.zig to generic riscv).

### CODE REVIEW

File: lib/std/Target.zig [hunk 1]
```diff
@@ -2002,6 +2001,8 @@ pub fn stackAlignment(target: Target) u16 {
         .wasm64,
         .loongarch64,
         => 16,
+        .riscv32,
+        => if (Target.riscv.featureSetHas(target.cpu.features, .e)) 4 else 16,
```
Comment (alexrp) [inline lib/std/Target.zig:2005]: For reference for other reviewers: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#ilp32e-calling-convention

File: src/codegen/llvm.zig [hunk 2]
```diff
@@ -553,7 +553,8 @@ const DataLayoutBuilder = struct {
             try self.typeAlignment(.float, 64, 64, 64, true, writer);
         }
         if (stack_abi != ptr_bit_width or self.target.cpu.arch == .msp430 or
```
Comment (alexrp) [inline src/codegen/llvm.zig:555]: Not related to your PR, but these checks look really brittle. Is there any reason we can't just always trust `std.Target.stackAlignment()`? And if there is, shouldn't that be fixed? There's also a special-case for CSKY up on line 396. Weird situation overall.

**Review by alexrp — APPROVED** [review]

**Comment by alexrp** [comment]
Note that you should be able to do something like `zig build -Dtarget=riscv32-freestanding -Dcpu=generic_rv32+e` (and similarly for other ISA extensions). An extra `Cpu` entry probably isn't necessary.

**Comment by alexrp** [comment]
> An extra `Cpu` entry probably isn't necessary.

That said, I do think there's an argument for adding `std.Target.Abi` tags for RV32E/RV64E (but maybe not until the specifications are actually finalized?). We need to do something similar for other ABI values (`ilp32`/`ilp32f`/`ilp32d` and `lp64`/`lp64f`/`lp64d`), and GCC/Clang [support](https://gcc.gnu.org/onlinedocs/gcc/RISC-V-Options.html) `ilp32e`/`lp64e` as `-mabi` arguments too. I intend to open an RFC issue on solving this soon-ish.

File: src/codegen/llvm.zig (hunk 2 above)
Comment (cheme) [reply src/codegen/llvm.zig:555]: I got the same feeling. Given current code looks like one would need a lot of variant target for riscv feature combination, which would not feel right either. Guess would need a bit of code design change to allow this.

**Comment by andrewrk** [comment]
Thanks for the patch, and thanks @alexrp for the additional review.

**Pull request merged**